universal-application-tool-0.0.1/app/controllers/admin/AdminProgramController.java

```diff
@@ -74,10 +83,23 @@ public Result update(Request request, long id) {
     Form<ProgramForm> programForm = formFactory.form(ProgramForm.class);
     ProgramForm program = programForm.bindFromRequest(request).get();
     try {
-      service.updateProgramDefinition(id, program.getName(), program.getDescription());
+      ErrorAnd<ProgramDefinition, ProgramServiceError> result =
+          service.updateProgramDefinition(id, program.getName(), program.getDescription());
+      if (result.isError()) {
+        String errorMessage = joinErrors(result.getErrors());
+        return ok(editView.render(request, id, program, errorMessage));
+      }
       return redirect(routes.AdminProgramController.index().url());
     } catch (ProgramNotFoundException e) {
       return notFound(String.format("Program ID %d not found.", id));
     }
   }
+
+  private String joinErrors(ImmutableSet<ProgramServiceError> errors) {
+    StringJoiner messageJoiner = new StringJoiner(". ", "", ".");
+    for (ProgramServiceError e : errors) {
+      messageJoiner.add(e.message());
+    }
+    return messageJoiner.toString();
+  }
```

Comment:
Suggestion (blocking): nearly identical code appears in [QuestionController](https://github.com/seattle-uat/civiform/blob/aab99bb53941afbb527ab2f68435e163d4fc2106/universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java#L68-L73). Recommend creating `class CiviFormController extends Controller` that implements `protected String joinErrors(...)` and make it a superclass of the controllers we define.

Reply:
Good idea! Done.